### ISSUE TITLE
[css-pseudo-4] `::nth-letter` pseudo-element

### ISSUE THREAD

**Issue by adactio** [w3c/csswg-drafts#3208]
I'd like to propose that there be an `::nth-letter` [pseudo-element](https://drafts.csswg.org/css-pseudo-4/) in CSS.

[I've written about it on my website](https://adactio.com/journal/14408).

Previous calls:

* [2003, Anne van Kesteren](https://annevankesteren.nl/2003/09/from-a-markover-to-pseudo-elements)
* [2011, Trent Walton](https://trentwalton.com/2011/01/18/controlling-web-typography/)
* [2011, Chris Coyier](https://css-tricks.com/a-call-for-nth-everything/)

[The last time this came up](https://github.com/w3c/csswg-drafts/issues/745), @frivoal said:

> This sits at the uncomfortable intersection of useful and freaking hard to implement.

But also:

> As for ::nth-letter, I am not aware of any substitute, but neither am I aware of strong use cases.

I would argue that the strong use cases are being demonstrated every day by any site using these JavaScript libraries:

* [Lettering.js](http://letteringjs.com/)
* [Splitting.js](https://splitting.js.org/)

Or these logo examples in the wild:

* [Patterns Day](https://patternsday.com/)
* [Ampersand](https://2018.ampersandconf.com/)

...along with most of [these variable font demos](https://codepen.io/mandymichael/).

Whether using HTML, or using JavaScript libraries, the only way to target specific letters is to wrap them in uneccessary elements. This can cause accessibility issues—a logo being read out letter-by-letter, for example, rather than as a whole word.

I understand that this would be hard to implement (given that `::first-letter` was hard enough) but I think the use-cases and accessibility benefits could outweigh the potential difficulty.

**Comment by tomhodgins** [comment]
I've published a demo earlier this week for `:nth-letter()` and `:nth-word()` pseudos, including a JavaScript implementation that can be tested/used from CSS: https://codepen.io/tomhodgins/pen/YJZyPr

**Comment by upsuper** [comment]
Yeah... `::first-letter` really taught us that it isn't something worth doing :) It should have been done in the `initial-letter` way...

**Comment by fantasai** [comment]
@upsuper `initial-letters` is a layout property that happens to apply to `:first-letter`, it doesn't magically select the first letter the way `text-combine-upright: digits` magically selects numbers ...

@tabatkins I don't think any reasonably marked up document would put an element boundary there, so it's OK for that to be handled as a simplified error of some kind. I think the basic point is that pseudo-elements of this type make implementers very unhappy* so a moderately convincing use case with an easy workaround is likely not enough motivation.

* A fun example is even-odd styling of :nth-letter() over a paragraph. Now insert a caret at character 5 and start typing. Somehow that has to work as expected, and reasonably preformantly as well because that's what the user expects. Other fun questions include “exactly how do Indic scripts work here?”, “what about Dutch ij?” (answer: probably not the same way as `:first-letter`), and other i18n inquiries.

**Comment by adactio** [comment]
@tabatkins said:

> One-off effects on top-level headings in a document still seem to be well-handled by just putting some spans around the letters you want to individually style.

Like I said, I knew it would be difficult to implement. Still, I'm surprised that difficulty of implementation would win out over the accessibility problems of wrapping individual letters in a bunch of `span`s.

> All of these examples (and I expect, nearly all reasonable examples you could dredge up beyond these)

Dredge up.

Interesting.

**Comment by tabatkins** [comment]
> Still, I'm surprised that difficulty of implementation would win out over the accessibility problems of wrapping individual letters in a bunch of spans.

The hierarchy of concerns tells us that certain players in the space matter more than others; user-a11y is generally more important than implementation difficulty. It is not, however, absolute - `::nth-letter()` has *significant* implementation difficulties, over and above even that of `::first-letter` which is already troublesome, *and* has less-common use-cases. All of these combine to at least make `::nth-letter()` a *rather low priority* project, versus many other things we could add to CSS.

Solving the a11y issues can be done more simply than adding `::nth-letter()`, such as having something in HTML or CSS that says "this element is not a semantic separation of its contained text from the surrounding text", which could be applied to the `span`s. This would also solve slightly wider problems than `::nth-letter()`, as anything that splits up a word for any reason could also use it, rather than the solution being limited solely to the use-case of applying different CSS to letter-based groupings.

> Dredge up.
>
> Interesting.

? I'm confused, it looks like you might be reading some deep meaning into my wording?

**Comment by shshaw** [comment]
Author of Splitting.js here. The difficulty of implementation is an issue I'm sure, but `::nth-letter` would allow for so many great possibilities  _especially_ if combined with a `letter-index()` counter and `letter-count()` total, [akin to this proposal to get the index of children](https://github.com/w3c/csswg-drafts/issues/1869). 

Bonus points if `::nth-word` could be implemented as well.

You can see some of the [really creative ways folks have utilized Splitting for text](https://codepen.io/collection/XpROaV), and it's not isolated to decorative headings. [More use cases](https://codepen.io/tag/splittext) can be found in GSAP with its [SplitText plugin](https://greensock.com/SplitText).

In the interim, I would love some proper accessibility support for splitting up text with `<span>`s, as mentioned. [I've tried several approaches in Splitting](https://github.com/shshaw/Splitting/issues/19) but can't find a right setup for keeping the text properly available to screenreaders. Perhaps `aria-role="inline"` that would treat an element as if it were not distinct from its siblings.

**Comment by AmeliaBR** [comment]
Worth mentioning that we've had lots of discussion in SVG about how to handle features that assign properties to individual characters (https://github.com/w3c/svgwg/issues/631). However, the conclusion we made there was to stict to predictable parsing (codepoints) over typographical elegance — an approach that isn't at all consistent with CSS `::first-letter`.

**Comment by faceless2** [comment]
See also https://github.com/w3c/svgwg/issues/537, and in particular the linked-to examples showing how complicated this is for some scripts, eg https://w3c.github.io/i18n-tests/quick-tests/svg-counting/svg-counting-001.

**Comment by Crissov** [comment]
Perhaps there should be a more extensive [snapshot profile](https://drafts.csswg.org/selectors-4/#profiles), i. e. as a separate specification, that also included properties or values, but as a selector, `::nth-letter()` would fit in very well with `:has()`.

**Comment by SelenIT** [comment]
Snapshot profile [is dead](https://github.com/w3c/csswg-drafts/issues/3925#issuecomment-492724267) (luckily), it's just the draft hasn't been updated accordingly yet (I'm going to make a relevant PR this weekend). IMO, there is already no problem with accessing any character of the text (as well as with selecting DOM elements basing on their contents) from JS APIs. It's _styling_ these things that CSS developers have demanded for years but can't do yet.

**Comment by shshaw** [comment]
Accessing characters from JS is problematic when you get into Emoji or
extended Unicode ranges where some “characters” are actually multiple, like
the “family” emojis 👨‍👩‍👧‍👦

The current solution is a messy RegEx pattern matching specific ranges of
characters (see some discussion here
https://github.com/shshaw/Splitting/issues/25).Having a built in RegEx
matched to target ligatures/combined characters Would be a huge help.

**Comment by SelenIT** [comment]
@shshaw well, `::first-letter` also has its problems with emojis (at least in Gecko), hope it's just a bug... And according to the [StackOverflow answer](https://stackoverflow.com/a/38901550/2533215) mentioned in the discussion you referenced, accessing the (unicode) character in position `n` in the string `string` appears to be possible with not too complicated functions like `(string, n) => string.match(new RegExp('(?<=.{' + n + '})', 'u'))[0]` or just `(string, n) => [...string][n]`.

**Comment by ByteEater-pl** [comment]
Thinking… Maybe that's a use case better suited for shadow DOM? If scripting is confined therein, authors reaching for reusable components needn't be concerned much. The text could be programatically split into text nodes in the light corresponding to grapheme clusters and `span`s inserted around each in the shadow. What remains is indexing – `::part` and `::theme` don't take arguments, maybe they should optionally to be resolved by script or some CSS variables in selectors machinery? Or is is Houdini's domain whereby an actual `nth-letter()` selector working in tandem with shadow DOM could be added (in which case I'd rather make it more general, e.g. `nth-specially-marked-item-in-container()`)?

**Comment by severdia** [comment]
Here's a use case for `::nth-letter() `(or at least second letter). When trying to reproduce historical texts like Shakespeare, initial letter doesn't work for several reasons:

1. There are hard line breaks in poetry and the dropcap should span two lines (or more in many cases) and it's not recognized (`<br />` doesn't work for semantic reasons)
2. Sometimes the dropcap is two letters (see below).
3. The next letter after the dropcap is often capitalized, but not always.

![css](https://user-images.githubusercontent.com/355924/115968923-000e2c80-a4ef-11eb-9958-d87c96ba219b.jpg)

If there's reason enough to support dropcaps, there should be reason enough to support them correctly, not superficially.

**Comment by faceless2** [comment]
You could do this with `initial-letter` to enclose the two "V" characters in a span - (or making it a "W" and just using an appropriate font...) So I'm not sure this is a case for `nth-letter`

However, here's an idea. How about we introduce control over which letters are considered part of the first-letter by using Unicode joiners? We could we specify that the `first-letter` pseudo-element must not break at a ZWJ and must break at ZWNJ.

That would allow us to support this example with `V&zwj;VHen to the sessions...`, and also cover quite a few of the edge cases that seem to be coming up with punctuation and first-letter, without any additional CSS properties. I suspect this would be pretty easy to implement too.

**Comment by severdia** [comment]
It's possible to ask font designers to include these kinds of examples as single characters, but not likely to happen.

I tried `initial-letter` with a `<span>` and it doesn't work with two letters, nor does it work on inline elements (only block elements, which breaks the dropcap formatting). Here's a Codepen: https://codepen.io/severdia/pen/BapMMzx

I think it's an interesting idea to control it with a Unicode word joiner. That should work for all the use cases I can think of. Thanks!

**Comment by johannesmutter** [comment]
>  As for ::nth-letter, I am not aware of any substitute, but neither am I aware of strong use cases.

A common use case for nth-letter/ nth-word / etc. are code or more general text editors and annotations for grammar, syntax, spelling, entities (NLP), etc. Which often require overlapping markup (see standoff properties / annotations for more). 

In above use cases styling the characters is insufficient: You will also want to register event listener, arbitrary data, … on it. So I’m not convinced an extension of the CSS spec would be a good idea.

Except for using canvas/ CSS Houdini, I currently don’t see a good alternative to using a million span tags that wrap each character. To keep the markup readable I'm storing the plain text string in a data attribute of e.g. a paragraph tag. _(however markup readability shouldn’t be a concern if you can inspect a readable JSON of the content)_
```html 
<p aria-description="Hello World">
  <span annotations={...}>H</span>
  <span annotations={...}>e</span>
  <span annotations={...}>l</span>
  <span annotations={...}>l</span>
  <span annotations={...}>o</span>
...
</p>
```


![standoff markup and annotations](https://user-images.githubusercontent.com/9489658/173823343-82ecfc98-0fbf-4857-9968-eeae97e2637d.png)

**Comment by kbabbitt** [comment]
While exploring staggered text fade-in animations, I built a rough prototype of `::nth-word()` in Blink and found the performance promising compared to wrapping each word in a `<span>`. I'd like to explore the details a bit more.

## Internationalization

There are existing solutions for determining what constitutes a "letter" (i.e. grapheme cluster) or "word," and I wouldn't want this feature to invent a worse wheel. There exists [precedent](https://drafts.csswg.org/css-text-3/#text-transform-mapping) in CSS to defer to the UA with a suggestion to use [UAX-29](https://www.unicode.org/reports/tr29/tr29-47.html) which is I believe what `Intl.Segmenter` uses. I think `::nth-letter()` and `::nth-word()` should follow suit.

## Counting words

Words can cross element boundaries, which would suggest that `::nth-word()` would be non-tree-abiding. Unfortunately, that forgoes a powerful capability: using counters, which when coupled with the `parse()` function (#12488) would give authors a lot of flexibility in determining how per-word effects progress. One might imagine a staggered text fade-in with specialized handling for various types of descendant blocks:

```css
@mixin --staggered {
    @starting-style { opacity: 0; }
    opacity: 1;
    transition: opacity 600ms;
    transition-delay: calc(6ms * parse(<number> counter(stagger)));
    counter-increment: stagger;
}
.fade-in {
    counter-reset: stagger;
}
.fade-in > p::nth-word(n) {
    @apply --staggered;
}
/* treat an inline code sample as an atomic unit */
.fade-in > pre.code {
    @apply --staggered;
}
/* stagger in gallery images one at a time */
.fade-in > .gallery {
    display: flex;
    flex-wrap: wrap;
    & > img {
        @apply --staggered;
    }
}
```

I would love if we could find a coherent way to make that work.

Barring that, we could imagine `word-index()` and `letter-index()` functions, akin to `sibling-index()`. Those would involve more baked-in indexing mechanics, including when indices reset. It would make it harder for authors to continue an effect like the above from one sibling paragraph to the next - they would need to keep a running word count to offset the delays in each paragraph. Or maybe there's a property that could be set to link indices together from one block to the next, and let blocks participate in indexing as atomic units, something like:

```css
word-indexing: [ restart | continue | block ]
```
Only has an effect on block-level elements. Word indexing always continues through inline-level elements.
* `restart` (initial value): the first word gets index 1. Each subsequent word's index increases by 1 until a descendant block is entered.
* `continue`: the first word gets index n+1, where n is the index of the last word prior to the block in in-order traversal.
* `block`: the block itself, and all words within it, are treated as if they are at index n+1, where n is defined as for `continue`.

**Comment by bramus** [comment]
> There exists [precedent](https://drafts.csswg.org/css-text-3/#text-transform-mapping) in CSS to defer to the UA with a suggestion to use [UAX-29](https://www.unicode.org/reports/tr29/tr29-47.html) which is I believe what `Intl.Segmenter` uses. I think `::nth-letter()` and `::nth-word()` should follow suit.

Yes.

> Words can cross element boundaries, which would suggest that `::nth-word()` would be non-tree-abiding. Unfortunately, that forgoes a powerful capability: using counters, which when coupled with the `parse()` function ([#12488](https://github.com/w3c/csswg-drafts/issues/12488)) would give authors a lot of flexibility in determining how per-word effects progress.

I agree that in general this needs to be solved, because a typical use-case here is having `<span>`/`<b>`/`<i>`/… elements nested inside the element whose words you want to count: more than often, you just want to count the words of the flattened text across all children.

If regular counters cannot be used _(why? Is it the inheritance?)_, what if we had a separate word counter, modeled after `counter`?

```css
.count-words-in-this-element {
  word-counter: <dashed-ident>;
}
.count-words-in-this-element::nth-word(n) {
  word-counter-increment: <dashed-ident>;
}
```

This would allow counting using an `<An+B>` pattern. For example to gradually fade out “the last 3 words” you could use `::nth-last-word(-n+3)` _(assuming we also do a `::nth-last-X` in addition to `::nth-X`)_. It would also allow nesting of word counters.

The value of the counter would be retrieved using a new `counter()` function that immediately returns a number, e.g. `counter-index()` — which would also immediately tackle requests like #1026

```css
.count-words-in-this-element::nth-word(n) {
  transition-delay: calc(
      6ms * counter-index(<dashed-ident>)
    );
}
```

_(Or `parse()`, of course … but that is more to type)_

> Or maybe there's a property that could be set to link indices together from one block to the next, and let blocks participate in indexing as atomic units

My fear here is that authors would end up writing the following CSS a lot just to make it work:

```css
.count-words-in-this-element * {
  word-indexing: continue;
}
```

I’m also not sure how this would play nice with nesting or to counting `<An+B>` patterns.

**Comment by kbabbitt** [comment]
> If regular counters cannot be used (why? Is it the inheritance?)

That's one issue, yes. Consider:
```html
<p>I made a tur<b>duck</b>en for the Big Game.</p>
```
```css
p {
  counter-reset: wordcount;
}
p::nth-word(n) {
  counter-increment: wordcount;
}
```
The word "turducken" generates three boxes. What values do these boxes get for `wordcount`? What if there's a rule that uses, or increments, `wordcount` on `<b>`?

**Comment by kbabbitt** [comment]
> what if we had a separate word counter, modeled after counter?

Yeah, this might be the way to go. I think having it count words rather than boxes is enough of a difference that it's reasonable to have a new notion of "counter." Or maybe there's a keyword we can add to `counter-reset` to designate a given counter as a word counter?

**Comment by bramus** [comment]
For reference, GSAP demos that would benefit from this: https://github.com/w3c/csswg-drafts/issues/13689#issuecomment-4101334716 _(the split text into characters section)_
